File: lib/src/models/render/render_info.dart
// A simple helper class to hold widget info
import 'dart:ui';
import 'package:json_annotation/json_annotation.dart';

part 'render_info.g.dart';

@JsonSerializable()
class RenderInfo {

  @OffsetConverter()
  final Offset position;

  @SizeConverter()
  final Size size;
  final double pixelRatio;

  RenderInfo({
    required this.position,
    required this.size,
    required this.pixelRatio,
  });

  factory RenderInfo.fromJson(Map<String, dynamic> json) =>_$RenderInfoFromJson(json);

  Map<String, dynamic> toJson() => _$RenderInfoToJson(this);

}


class OffsetConverter implements JsonConverter<Offset, Map<String, dynamic>> {
  const OffsetConverter();

  @override
  Offset fromJson(Map<String, dynamic> json) {
    return Offset(
      (json['dx'] as num).toDouble(),
      (json['dy'] as num).toDouble(),
    );
  }

  @override
  Map<String, dynamic> toJson(Offset offset) => {
        'dx': offset.dx,
        'dy': offset.dy,
      };
}

class SizeConverter implements JsonConverter<Size, Map<String, dynamic>> {
  const SizeConverter();

  @override
  Size fromJson(Map<String, dynamic> json) {
    return Size(
      (json['width'] as num).toDouble(),
      (json['height'] as num).toDouble(),
    );
  }

  @override
  Map<String, dynamic> toJson(Size size) => {
        'width': size.width,
        'height': size.height,
      };
}

File: android/src/main/kotlin/com/example/flutter_sceneview/models/render/RenderInfo.kt
package com.example.flutter_sceneview.models.render

data class RenderInfo(
    val position: Offset, val size: Size, val pixelRatio: Double
) {
    companion object {
        fun fromMap(map: Map<*, *>): RenderInfo? {
            val positionMap = map["position"] as? Map<*, *>
            val sizeMap = map["size"] as? Map<*, *>

            return if (positionMap != null && sizeMap != null) {
                RenderInfo(
                    position = Offset(
                        (positionMap["dx"] as Number).toDouble(),
                        (positionMap["dy"] as Number).toDouble(),
                        0.0
                    ), size = Size(
                        (sizeMap["width"] as Number).toDouble(),
                        (sizeMap["height"] as Number).toDouble()
                    ), pixelRatio = (map["pixelRatio"] as Number).toDouble()
                )
            } else null
        }
    }
}

data class Offset(val dx: Double, val dy: Double, val dz: Double = 0.0)
data class Size(val width: Double, val height: Double)